XenModel/Actions/Updates/DownloadAndUpdateClientAction.cs

```diff
@@ -181,35 +183,56 @@ protected override void Run()
                 if (Cancelling)
                     throw new CancelledException();
             }
+        }
 
-            ValidateMsi();
-
-            if (!File.Exists(outputPathAndFileName))
+        public void InstallMsi()
+        {
+            if (!File.Exists(_outputPathAndFileName))
                 throw new Exception(Messages.DOWNLOAD_CLIENT_INSTALLER_MSI_NOT_FOUND);
 
+            // Install the msi            
+            try
+            {
+                ValidateMsi();

```

Comment:
I don't think it's a good idea to move the validation outside the action. For one thing  if it fails, the exception will be propagated upwards and caught by xencenter as an unhandled exception without showing a meaningful error.